### PR TITLE
propagation rule and test added for broadcast op

### PR DESCRIPTION
**Pull request by jjsjann123**
Stacked PRs:
#1755 enabling layout propagation through runtime
==== #1792 propagation rule for broadcast **_<- this one_**
#1790 propagation rule for binary op
#1788 adding layout inference pass

What's in this PR:

BroadcastOp propagation tries to push all new broadcast iterdomain as outer dimensions for the output tensor.

- [x] adding propagate rule for broadcast op;
- [x] adding cpp test to verify propagation rule;

### CODE REVIEW

**Comment by jjsjann123** [comment]
!build!

File: csrc/preseg_passes/allocation_order_inference.cpp [hunk 1]
```diff
@@ -77,6 +77,72 @@ void AllocationOrderInferencer::handle(UnaryOp* op) {
   propagateAllocationOrder(in, out);
 }
 
+// BroadcastOp propagation:
```
Comment (jjsjann123) [inline csrc/preseg_passes/allocation_order_inference.cpp:80]: note to @zasdfgbnm 

I kept the old implementation here, i.e. pushing all broadcast iterdomain to outer. I'm imagining this allows us to have better allocation order, i.e. more iter domains staying closer to each other, so better vectorization?!

File: csrc/preseg_passes/allocation_order_inference.cpp [hunk 2]
```diff
@@ -77,6 +77,72 @@ void AllocationOrderInferencer::handle(UnaryOp* op) {
...
+//
+//   note 2:
+//       computing the new output allocation order
+//       We'll scan through the rfactor domain of output where:
+//       a. insert any broadcast iterdomain index as we encounter them
+//       b. adjust the index of entry from input's rfactor domain
+//
+//   so output TV1 will have stride order {1, 4, 0, 3, 2}
+void AllocationOrderInferencer::handle(BroadcastOp* op) {
+  TensorView* out = dynamic_cast<TensorView*>(op->out());
```
Comment (wujingyue) [inline csrc/preseg_passes/allocation_order_inference.cpp:110]:
```suggestion
  auto* out = dynamic_cast<TensorView*>(op->out());
```

`dynamic_cast<TensorView*>` already spells out the type, so using `auto*` is fine and encouraged.

File: csrc/preseg_passes/allocation_order_inference.cpp [hunk 3]
```diff
@@ -77,6 +77,72 @@ void AllocationOrderInferencer::handle(UnaryOp* op) {
...
+  if (out == nullptr) {
+    return;
+  }
+  TensorView* in = op->in()->as<TensorView>();
```
Comment (wujingyue) [inline csrc/preseg_passes/allocation_order_inference.cpp:114]:
```suggestion
  auto* in = op->in()->as<TensorView>();
```

File: csrc/preseg_passes/allocation_order_inference.cpp [hunk 4]
```diff
@@ -77,6 +77,72 @@ void AllocationOrderInferencer::handle(UnaryOp* op) {
...
+  if (const auto& iter = alloc_order_map_.find(in);
+      iter != alloc_order_map_.end()) {
```
Comment (wujingyue) [inline csrc/preseg_passes/allocation_order_inference.cpp:116]: Can you do an early return to reduce indentation? 

```
auto iter = alloc_order_map_.find(in);
if (iter == alloc_order_map.end()) {
  return;
}
```

File: csrc/preseg_passes/allocation_order_inference.cpp [hunk 5]
```diff
@@ -77,6 +77,72 @@ void AllocationOrderInferencer::handle(UnaryOp* op) {
...
+    AllocationOrder out_order;
+    int64_t out_rank = static_cast<int64_t>(out->nDims());
```
Comment (wujingyue) [inline csrc/preseg_passes/allocation_order_inference.cpp:118]:
```suggestion
    auto out_rank = static_cast<int64_t>(out->nDims());
```

File: csrc/preseg_passes/allocation_order_inference.cpp [hunk 6]
```diff
@@ -77,6 +77,72 @@ void AllocationOrderInferencer::handle(UnaryOp* op) {
...
+//       stack all broadcast in rfactor [b3, b4].
+//       concat([b3, b4], [i0, i2, i1])
+//
+//   note 2:
+//       computing the new output allocation order
+//       We'll scan through the rfactor domain of output where:
+//       a. insert any broadcast iterdomain index as we encounter them
+//       b. adjust the index of entry from input's rfactor domain
+//
+//   so output TV1 will have stride order {1, 4, 0, 3, 2}
```
Comment (wujingyue) [inline csrc/preseg_passes/allocation_order_inference.cpp:108]:
```suggestion
//   so output TV1 will have allocation order {1, 4, 0, 3, 2}
```

File: csrc/preseg_passes/allocation_order_inference.cpp [hunk 7]
```diff
@@ -77,6 +77,72 @@ void AllocationOrderInferencer::handle(UnaryOp* op) {
   propagateAllocationOrder(in, out);
 }
 
+// BroadcastOp propagation:
+//   1. preserves all allocation order of input iterdomain;
+//   2. stacks all added broadcast iter domain on outputs as outer dimensions in
+//   their natural position
+//
+// e.g.
+//   TV0 rfactor dom [i0, i1, i2] @ stride order {0, 2, 1}
```
Comment (wujingyue) [inline csrc/preseg_passes/allocation_order_inference.cpp:86]:
```suggestion
//   TV0 rfactor dom [i0, i1, i2] @ allocation order {0, 2, 1}
```

File: csrc/preseg_passes/allocation_order_inference.cpp [hunk 8]
```diff
@@ -77,6 +77,72 @@ void AllocationOrderInferencer::handle(UnaryOp* op) {
...
+        // see note 2.b
+        offset_table[offset_entry++] = broadcast_seen_so_far;
+      }
+    }
+
+    for (auto i : c10::irange(in->nDims())) {
+      auto order_entry = iter->second[i];
+      out_order.push_back(order_entry + offset_table[order_entry]);
+    }
+    alloc_order_map_[out] = out_order;
```
Comment (wujingyue) [inline csrc/preseg_passes/allocation_order_inference.cpp:142]: Nit: It took me 10 minutes to understand this :) I'd do the following: 
1. apply in's alloc_order to in's rfactor => {i0, i2, i1}
2. map {i0, i2, i1} to out's root => {i0', i2', i1'} using PairwiseRootMap. 
3. prepend broadcast dimensions => {b3, b4, i0', i2', i1'}
4. computePermutation(out's rfactor, {b3, b4, i0', i2', i1'}) => {1,4,0,3,2}

**Review by wujingyue — APPROVED** [review]

File: csrc/preseg_passes/allocation_order_inference.cpp [hunk 9]
```diff
@@ -56,14 +57,88 @@ class AllocationOrderInferencer : public IterVisitor {
...
+  auto in_to_out_map = PairwiseRootDomainMap(in, out).mapProducerToConsumer();
+  const auto& in_root_domain =
+      TensorDomain::noReductions(in->getMaybeRFactorDomain());
+
+  // step 2: push each mapped iterdomain
+  for (auto index : iter->second) {
+    alloc_domain.push_back(in_to_out_map.at(in_root_domain.at(index)));
+  }
+
+  // step 3: compute permutatoin
```
Comment (jacobhinkle) [inline csrc/preseg_passes/allocation_order_inference.cpp:132]:
```suggestion
  // step 3: compute permutation
```

File: csrc/preseg_passes/allocation_order_inference.cpp [hunk 10]
```diff
@@ -56,14 +57,88 @@ class AllocationOrderInferencer : public IterVisitor {
...
+  std::optional<AllocationOrder> permutation = ir_utils::computePermutation(
+      TensorDomain::noReductions(out->getMaybeRFactorDomain()), alloc_domain);
```
Comment (jacobhinkle) [inline csrc/preseg_passes/allocation_order_inference.cpp:134]: We should not need `noReductions` here since `out` is the output of a `BroadcastOp`.

**Comment by jjsjann123** [comment]
!build

**Comment by jjsjann123** [comment]
Tests passed locally. I'll merge once CI build finishes.

**Pull request merged**